components/brave_wallet/browser/meld_integration_service.cc

```diff
@@ -0,0 +1,549 @@
+                               IDS_WALLET_PARSING_ERROR)});
+    return;
+  }
+  std::move(reply_callback).Run(std::move(payment_methods), std::nullopt);
+}
+
+// static
+GURL MeldIntegrationService::GetFiatCurrenciesURL(mojom::MeldFilterPtr filter) {
+  return AppendFilterParams(
+      GURL(base::StringPrintf("%s/service-providers/properties/fiat-currencies",
+                              GetMeldAssetRatioBaseURL().c_str())),
+      std::move(filter),
+      base::flat_map<std::string, std::string>{
+          {"includeServiceProviderDetails", "false"},
+          {"accountFilter", "false"}});
+}
+
+void MeldIntegrationService::GetFiatCurrencies(
+    mojom::MeldFilterPtr filter,
+    GetFiatCurrenciesCallback callback) {
+  auto internal_callback =
+      base::BindOnce(&MeldIntegrationService::OnGetFiatCurrencies,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+
+  auto conversion_callback = base::BindOnce(&SanitizeJson);
+  api_request_helper_->Request(
+      "GET", GetFiatCurrenciesURL(std::move(filter)), "", "",
+      std::move(internal_callback), MakeMeldApiHeaders(),
+      {.auto_retry_on_network_change = true, .enable_cache = true},
+      std::move(conversion_callback));
+}
+
+void MeldIntegrationService::OnGetFiatCurrencies(
+    GetFiatCurrenciesCallback callback,
+    APIRequestResult api_request_result) const {
+  if (!api_request_result.Is2XXResponseCode() &&
+      !NeedsToParseResponse(api_request_result.response_code())) {
+    std::move(callback).Run(
+        std::nullopt, std::vector<std::string>{
+                          l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR)});
+    return;
+  }
+
+  if (auto errors = ParseMeldErrorResponse(api_request_result.value_body());
+      errors) {
+    std::move(callback).Run(std::nullopt, errors);
+    return;
+  }
+
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE,
+      base::BindOnce(&ParseFiatCurrencies, api_request_result.TakeBody()),
+      base::BindOnce(&MeldIntegrationService::OnParseFiatCurrencies,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void MeldIntegrationService::OnParseFiatCurrencies(
+    GetFiatCurrenciesCallback reply_callback,
+    std::optional<std::vector<mojom::MeldFiatCurrencyPtr>> fiat_currencies)
+    const {
+  if (!fiat_currencies) {
+    std::move(reply_callback)
+        .Run(std::nullopt, std::vector<std::string>{l10n_util::GetStringUTF8(
+                               IDS_WALLET_PARSING_ERROR)});
+    return;
+  }
+  std::move(reply_callback).Run(std::move(fiat_currencies), std::nullopt);
+}
+
+// static
+GURL MeldIntegrationService::GetCryptoCurrenciesURL(
+    mojom::MeldFilterPtr filter) {
+  return AppendFilterParams(
+      GURL(base::StringPrintf(
+          "%s/service-providers/properties/crypto-currencies",
+          GetMeldAssetRatioBaseURL().c_str())),
+      std::move(filter),
```

Comment:
AppendFilterParams does not "consume" filter, so std::move() and AppendFilterParams should accept it by const &